### ISSUE TITLE
Problem with Encoding

### ISSUE THREAD

**Issue by oriongame** [BearishSun/BansheeEngine#43]
In Source/BansheeUtility/Source/BsString.cpp:131 and Source/BansheeUtility/Source/BsString.cpp:388
Your engine use dont good converter wchar <-> char.

```
    WString toWString(const String& source)
    {
        return WString(source.begin(), source.end());
    }

    String toString(const WString& source)
    {
        return String(source.begin(), source.end());
    }
```

If use russian encoding, the engine don't see path.

Right to use:

```

#define TSTR   template < typename TypeVarStr>
#define TDSTR  template < typename TStr1, typename TStr2>

    void WCharToChar(char* str1, const wchar_t* wstr2, size_t max) {
        wcstombs(str1, wstr2, max);
    }

    void CharToWChar(wchar_t* wstr1, const char* str2, size_t max) {
        mbstowcs(wstr1, str2, max);
    }

    TSTR int32_t inline StrCmp(const TypeVarStr *str1, const TypeVarStr *str2) {

        while ((*str1 != '\0') && (*str1 == *str2)) {
            str1++;
            str2++;
        }

        return (TypeVarStr)*str1 - (TypeVarStr)*str2;
    }

    TSTR int32_t inline StrCmpCopy(const TypeVarStr *str1, const TypeVarStr *str2) {

        while ((*str1 != '\0') && (*str1 == *str2)) {
            str1++;
            str2++;
        }

        return (TypeVarStr)*str1 - (TypeVarStr)*str2;
    }

    TSTR uint32_t inline StrLen(TypeVarStr *str1) {
        uint32_t len = 0;
        while (*str1 != 0x00) {
            str1++;
            len++;
        }
        return len;
    }

    TSTR void inline StrCpy(TypeVarStr *str1, const TypeVarStr *str2) {
        TypeVarStr *TmpStr = str1;
        while ((*TmpStr++ = *str2++) != 0x00);
    }

    TSTR void inline StrCat(TypeVarStr *str1, const TypeVarStr *str2)
    {
        StrCpy(&str1[StrLen(str1)], str2);
    }

    TSTR void inline StrClr(TypeVarStr *Str1, uint32_t Size) {
        for (uint32_t i = 0; i < Size; i++)
            Str1[i] = '\0';
    }

    TDSTR void ConvertStr(TStr1* str1, const TStr2* str2) {

        if ((StrCmp(typeid(str1).name(), "char *") == 0) && (StrCmp(typeid(str2).name(), "const char *") == 0)) return;
        if ((StrCmp(typeid(str1).name(), "wchar_t *") == 0) && (StrCmp(typeid(str2).name(), "const wchar_t *") == 0)) return;
        if ((StrCmp(typeid(str1).name(), "char * __ptr64") == 0) && (StrCmp(typeid(str2).name(), "char * __ptr64") == 0)) return;
        if ((StrCmp(typeid(str1).name(), "wchar_t * __ptr64") == 0) && (StrCmp(typeid(str2).name(), "const wchar_t * __ptr64") == 0)) return;

        if ((StrCmp(typeid(str1).name(), "char *") == 0) || (StrCmp(typeid(str1).name(), "char * __ptr64") == 0))
            WCharToChar((char *)str1, (wchar_t*)str2, StrLen(str2));
        else
            CharToWChar((wchar_t*)str1, (char *)str2, StrLen(str2));
    }

    TDSTR void inline StrCpy(TStr1 *str1, const TStr2 *str2) {

        uint32_t Len = StrLen(str2);
        if (Len == 0) return;
        TStr1 *tmp = new TStr1[Len + 1];
        ConvertStr(tmp, str2);
        tmp[Len] = NULL;
        StrCpy<TStr1>(str1, tmp);
        str1[Len] = NULL;
        delete[] tmp;
    }

    TDSTR void inline StrCat(TStr1 *str1, const TStr2 *str2) {
        uint32_t Len = StrLen(str2);
        if (Len == 0) return;
        TStr1 *tmp = new TStr1[Len + 1];
        ConvertStr(tmp, str2);
        tmp[Len] = NULL;
        StrCat<TStr1>(str1, tmp);
        delete[] tmp;
    }

    TDSTR int32_t inline StrCmp(TStr1 *str1, const TStr2 *str2) {
        uint32_t Len = StrLen(str2);
        if (Len == 0) return -1;
        TStr1 *tmp = new TStr1[Len + 1];
        tmp[Len] = NULL;
        StrCpy(tmp, str2);
        //ConvertStr(tmp, str2);
        int32_t Result = StrCmpCopy<TStr1>(str1, tmp);
        delete[] tmp;
        return Result;
    }

    WString toWString(const String& source)
    {
        UINT32 SizeString = StrLen(source.c_str());
        if (SizeString == 0) return L"";

        wchar_t   *Str = new wchar_t[SizeString + 1];
        Str[SizeString] = 0;

        ConvertStr(Str, source.c_str());

        WString ResultString(Str);


        delete[] Str;

        return ResultString;
        //return WString(source.begin(), source.end());
    }

    String toString(const WString& source)
    {
        UINT32 SizeString = StrLen(source.c_str());
        if (SizeString == 0) return "";

         char   *Str = new char[SizeString + 1];
         Str[SizeString] = 0;

         ConvertStr(Str, source.c_str());

         String ResultString(Str);


         delete[] Str;

        return ResultString;

        //return String(source.begin(), source.end());
    }

```

and in main function write setlocale(0, "RUSSIAN");
Or if you love use std, rewrite by itself this example:

```

#include <iostream>
#include <locale>
#include <string>
#include <vector>

std::wstring wide_string(std::string const &s, std::locale const &loc)
{
    if (s.empty())
        return std::wstring();
    std::ctype<wchar_t> const &facet = std::use_facet<std::ctype<wchar_t> >(loc);
    char const *first = s.c_str();
    char const *last = first + s.size();
    std::vector<wchar_t> result(s.size());

    facet.widen(first, last, &result[0]);

    return std::wstring(result.begin(), result.end());
}

std::string narrow_string(std::wstring const &s, std::locale const &loc, char default_char = '?')
{
    if (s.empty())
        return std::string();
    std::ctype<wchar_t> const &facet = std::use_facet<std::ctype<wchar_t> >(loc);
    wchar_t const *first = s.c_str();
    wchar_t const *last = first + s.size();
    std::vector<char> result(s.size());

    facet.narrow(first, last, default_char, &result[0]);

    return std::string(result.begin(), result.end());
}

char const *russian_locale_designator = "rus";

int main()
{
    std::locale loc(russian_locale_designator);

    std::wstring const s = L"русский текст";
    std::string  const sa = ::narrow_string(s, loc);
    std::wstring const sw = ::wide_string(sa, loc);

    std::locale::global(loc);

    std::wcout << "original  wide: " << s << std::endl;
    std::cout << "wide -> narrow: " << sa << std::endl;
    std::wcout << "narrow -> wide: " << sw << std::endl;

    std::system("PAUSE");
}

```

good luck!

**Comment by BearishSun** [comment]
Hi, thanks for your input. 

Is there any reason why wide strings are not suitable in this case? Path is internally always stored as a wide Unicode string, and you can use Path::toWString() to get it. Non-ASCII narrow strings could be useful, but don't solve the problem for larger character sets like Japanese or Chinese text.

**Comment by oriongame** [comment]
> Is there any reason why wide strings are not suitable in this case? Path is internally always stored as a wide Unicode string, and you can use Path::toWString() to get it. Non-ASCII narrow strings could be useful, but don't solve the problem for larger character sets like Japanese or Chinese text.
Your engine use more library. Example lib for font FreeType lib.
This is lib have method for open file only ASCII.
```
  FT_New_Face( FT_Library   library,
               const char*  filepathname,
               FT_Long      face_index,
               FT_Face     *aface );
```
 Path::toWString()  must convert in char for filepathname.

```
		FT_Face face;
		error = FT_New_Face(library, filePath.toString().c_str(), 0, &face);
```
If start client on russian PC:

```
Fatal error occurred and the program has to terminate!
InternalErrorException - Failed to load font file: C:\Users\!5@6\Desktop\BansheeEngine-master\Source\BansheeEditorExec\Data\Raw\Engine\arial.ttf. Unknown error.
in class std::shared_ptr __thiscall BansheeEngine::FontImporter::import(const class BansheeEngine::Path &,class std::shared_ptr) [C:\Users\Серж\Desktop\BansheeEngine-master\Source\BansheeFontImporter\Source\BsFontImporter.cpp:72]
```
See path?

> !5@6

Its name user in Russian encoding in ASCII, bad encoding, because name user "Серж".
And this problem there is lib OAAudio lib, see debug example:
http://c2n.me/3BQy6eA

**Comment by BearishSun** [comment]
That makes sense, it will be fixed, thanks.

**Comment by BearishSun** [comment]
Path encoding is now fixed with 8fb09fee8b598473682e6592d86e461ca0efc9f2.
Better Unicode encoding conversion for strings is available with 8a243ae52e48d7eeffe892a82c70b3320a3005b1.
